common/src/main/java/io/druid/collections/LoadBalancingPool.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.collections;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Supplier;
+import com.metamx.common.logger.Logger;
+
+import java.io.IOException;
+import java.util.Queue;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Simple load balancing pool that always returns the least used resource
+ *
+ * @param <T>
+ */
+public class LoadBalancingPool<T> implements Supplier<ResourceHolder<T>>
+{
+  private static final Logger log = new Logger(LoadBalancingPool.class);
+
+  private final Supplier<T> generator;
+
+  private final Queue<CountingHolder> objects;
+
+  private final int maxCapacity;
+  private final AtomicInteger size = new AtomicInteger(0);
+
+  public LoadBalancingPool(int maxCapacity, Supplier<T> generator)
+  {
+    Preconditions.checkArgument(maxCapacity > 0, "capacity must be greater than 0");
+    Preconditions.checkNotNull(generator);
+
+    this.generator = generator;
+    this.maxCapacity = maxCapacity;
+    this.objects = new PriorityBlockingQueue<>(maxCapacity);
+  }
+
+  public ResourceHolder<T> get()
```

Comment:
IMO it would be simpler and also OK to create all the things up front rather than scaling them up as they are needed. This is because they aren't getting scaled back down, so they should hit the max and stay there pretty soon anyway.